File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.py
import sys
input = sys.stdin.readline


class TrieNode:
    def __init__(self):
        self.children = dict()
        self.value = None


class Trie:
    def __init__(self):
        self.root = TrieNode()


    def insert(self, number):
        chk = self.root
        flag = True

        for i in range(len(number)):
            if number[i] in chk.children:
                chk = chk.children[number[i]]
                if hasattr(chk, 'last'): flag = False
                if i == len(number) - 1: flag = False
            else:
                new = TrieNode()
                new.value = number[i]
                chk.children[number[i]] = new
                chk = chk.children[number[i]]
            if i == len(number) - 1:
                chk.last = True

        return flag


t = int(input())

for i in range(t):
    n = int(input())
    trie = Trie()
    result = True

    for j in range(n):
        number = input().rstrip()
        flag = trie.insert(number)
        if not flag: result = False

    print('YES' if result else 'NO')

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.cc
#include <iostream>
#include <string>

using namespace std;

class Node {
 public:
  Node* children[10];
  bool last;

  Node() {
    last = false;
    for (int i = 0; i < 10; i++) {
      children[i] = nullptr;
    }
  }
};

class Trie {
  Node* root;

 public:
  Trie() { root = new Node(); }

  bool insert(string target) {
    Node* node = root;
    bool result = true;

    for (int i = 0; i < target.length(); i++) {
      if (node->children[target[i] - '0']) {
        node = node->children[target[i] - '0'];
        if (node->last || i == target.length() - 1) result = false;
      } else {
        node->children[target[i] - '0'] = new Node();
        node = node->children[target[i] - '0'];
      }
    }

    node->last = true;
    return result;
  }
};

int main() {
  int t;
  cin >> t;

  for (int i = 0; i < t; i++) {
    int n;
    bool result = true;
    Trie trie;

    cin >> n;

    for (int j = 0; j < n; j++) {
      string num;
      cin >> num;

      if (!trie.insert(num)) result = false;
    }

    if (result)
      cout << "YES" << endl;
    else
      cout << "NO" << endl;
  }

  return 0;
}